trixie: add psql and venv migrations

kay0u opened the pull request:
## The problem

https://github.com/YunoHost-Apps/immich_ynh/issues/322

## Solution

...

## PR Status

Not tested, I accidentally dropped all my psql db lul

## TODO Postgresql
- [x] Add rebuild index
- [ ] Test migration

## TODO Python
- [ ] Use `python -m venv --upgrade VENV_DIR` instead of rebuilding the venv
- [ ] Test on not yet supported python apps


## How to test

...

kay0u committed:
migrations: Add migration script for upgrading PostgreSQL from 15 to 17

kay0u committed:
locales: Add migration description strings

alexAubin commented on src/migrations/0038_rebuild_python_venv_in_trixie.py:
```diff
@@ -0,0 +1,207 @@
+
+
+class MyMigration(Migration):
+    """
+    After the update, recreate a python virtual env based on the previously
+    generated requirements file
+    """
+
+    ignored_python_apps = [
+        "diacamma",  # Does an ugly sed in the sites-packages/django_auth_ldap3_ad
+        "homeassistant",  # uses a custom version of Python
+        "immich",  # uses a custom version of Python
+        "kresus",  # uses virtualenv instead of venv, with --system-site-packages (?)
+        "librephotos",  # runs a setup.py ? not sure pip freeze / pip install -r requirements.txt is gonna be equivalent ..
+        "mautrix",  # install stuff from a .tar.gz
+        "microblogpub",  # uses poetry ? x_x
+        "mopidy",  # applies a custom patch?
+        "motioneye",  # install stuff from a .tar.gz
+        "pgadmin",  # bunch of manual patches
+        "searxng",  # uses --system-site-packages ?
+        "synapse",  # specific stuff for ARM to prevent local compiling etc
+        "matrix-synapse",  # synapse is actually installed in /opt/yunohost/matrix-synapse because ... yeah ...
+        "tracim",  # pip install -e .
+        "weblate",  # weblate settings are .. inside the venv T_T
+    ]
```
Zmlerf theoretically we should re-review all the app catalog by grepping apps which seem to use a virtualenv and check which apps are exception that can't be properly rebuilt because they use non-standard stuff

kay0u replied:
After a quick review of all apps with a venv: 
- [searxngs_ynh](https://github.com/YunoHost-Apps/searxng_ynh/blob/master/scripts/_common.sh#L31) seems ok now
- [paperless-ngx_ynh](https://github.com/YunoHost-Apps/paperless-ngx_ynh/blob/master/scripts/install#L42) I'm not sure, could work
- [mopidy](https://github.com/YunoHost-Apps/mopidy_ynh/blob/master/scripts/_common.sh#L13) seems ok
- matrix-synapse: I can't find this app
---
- [diacamma_ynh](https://github.com/YunoHost-Apps/diacamma_ynh/blob/master/scripts/_common.sh#L42) Does an ugly sed
- [django-for-runners_ynh](https://github.com/YunoHost-Apps/django-for-runners_ynh/blob/master/scripts/_common.sh#L94) pip-sync is used, I'm not sure if it's a problem
- [django-fritzconnection_ynh](https://github.com/YunoHost-Apps/django-fritzconnection_ynh/blob/master/scripts/_common.sh#L94) same, pip-sync
- [funkwhale_ynh](https://github.com/YunoHost-Apps/funkwhale_ynh/blob/master/scripts/upgrade#L64) install from a folder ./api?
- [homeassistant_ynh](https://github.com/YunoHost-Apps/homeassistant_ynh/blob/master/scripts/_common.sh#L47) uv is used
- [immich_ynh](https://github.com/YunoHost-Apps/immich_ynh/blob/master/scripts/_common.sh#L198) uv is used
- [indico_ynh](https://github.com/YunoHost-Apps/indico_ynh/blob/master/scripts/install#L76) symlink between venv and static web pages
- [kresus](https://github.com/YunoHost-Apps/kresus_ynh/blob/master/scripts/install#L46) uses virtualenv
- [motioneye](https://github.com/YunoHost-Apps/motioneye_ynh/blob/master/scripts/install#L46) install from .tar.gz
- [lasuite-docs_ynh](https://github.com/YunoHost-Apps/lasuite-docs_ynh/blob/main/scripts/install#L94) moving stuff into the venv
- [librephotos_ynh](https://github.com/YunoHost-Apps/librephotos_ynh/blob/master/scripts/_common.sh#L101) setup.py install
- [mautrix_telegram_ynh](https://github.com/YunoHost-Apps/mautrix_telegram_ynh/blob/master/scripts/install#L113) install from .tar.gz
- [microblogpub_ynh](https://github.com/YunoHost-Apps/microblogpub_ynh/blob/master/scripts/install#L47) uses poetry
- [pgadmin](https://github.com/YunoHost-Apps/pgadmin_ynh/blob/master/scripts/_common.sh#L56) bunch of manual patches
- [pretalx_ynh](https://github.com/YunoHost-Apps/pretalx_ynh/blob/master/scripts/install#L73) ynh_replace into the venv
- [synapse_ynh](https://github.com/YunoHost-Apps/synapse_ynh/blob/master/scripts/_common.sh#L34) ynh_setup_source into the venv
- [syncserver-rs_ynh](https://github.com/YunoHost-Apps/syncserver-rs_ynh/blob/master/scripts/install#L80) uses poetry
- [tracim](https://github.com/YunoHost-Apps/tracim_ynh/blob/master/scripts/install#L87) pip install .
- [weblate_ynh](https://github.com/YunoHost-Apps/weblate_ynh/blob/master/scripts/install#L118) settings are .. inside the venv T_T
- [yunohost_appgenerator_ynh](https://github.com/YunoHost-Apps/yunohost_appgenerator_ynh/blob/master/scripts/install#L27) uses pdm

tituspijean replied:
As of 2026-08-05, the only change to this list is `syncserver-rs_ynh` which does not use `poetry` anymore, though it uses `uv` but the resulting `venv` seems cleanly used.

kay0u commented on src/migrations/0037_postgresql_15_to_17.py:
```diff
@@ -0,0 +1,115 @@
+        if not self.package_is_installed(f"postgresql-{TARGET_VERSION}"):
+            raise YunohostValidationError(f"migration_{MIGRATION_NUMBER:04d}_postgresql_{TARGET_VERSION}_not_installed")
+
+        # Make sure there's a 15 cluster
```
```suggestion
        # Make sure there's a "PREVIOUS_VERSION" cluster
```

Thovi98 commented on src/migrations/0037_postgresql_15_to_17.py:
```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/env python3
+#
+# Copyright (c) 2024 YunoHost Contributors
```
```suggestion
# Copyright (c) 2025 YunoHost Contributors
```
I don’t know maybe those dates are automatically updated on push?

Josue-T commented:
Probably related to https://github.com/YunoHost/issues/issues/2737

ewilly commented:
@kay0u some comments about the postgresql migration:
* I had to modify the file `/etc/postgresql-common/user_clusters` from `* * 15 main postgres` to `* * 17 main postgres`
* I had to fix the collation mismatch with that bash command:
```
databases=$(LC_ALL=C sudo --login --user=postgres PGUSER=postgres PGPASSWORD="$(cat /etc/yunohost/psql)" psql --tuples-only --no-align --dbname="postgres" --command="SELECT datname FROM pg_database WHERE datistemplate = false OR datname = 'template1';")
for db in $databases
do
  LC_ALL=C sudo --login --user=postgres PGUSER=postgres PGPASSWORD="$(cat /etc/yunohost/psql)" psql --dbname="$db" --command="REINDEX DATABASE $db;"
  LC_ALL=C sudo --login --user=postgres PGUSER=postgres PGPASSWORD="$(cat /etc/yunohost/psql)" psql --dbname="$db" --command="ALTER DATABASE $db REFRESH COLLATION VERSION;"
done
```
sorry I'm not able to translate that in python language

Salamandar commented:
I guess let's first split the actual code from the bookworm migrations and reuse it in the trixie migrations?

Salamandar commented:
See #2274

kay0u commented:
> @kay0u some comments about the postgresql migration:
> 
>     * I had to modify the file `/etc/postgresql-common/user_clusters` from `* * 15 main postgres` to `* * 17 main postgres`
> 
>     * I had to fix the collation mismatch with that bash command:
> 
> 
> ```
> databases=$(LC_ALL=C sudo --login --user=postgres PGUSER=postgres PGPASSWORD="$(cat /etc/yunohost/psql)" psql --tuples-only --no-align --dbname="postgres" --command="SELECT datname FROM pg_database WHERE datistemplate = false OR datname = 'template1';")
> for db in $databases
> do
>   LC_ALL=C sudo --login --user=postgres PGUSER=postgres PGPASSWORD="$(cat /etc/yunohost/psql)" psql --dbname="$db" --command="REINDEX DATABASE $db;"
>   LC_ALL=C sudo --login --user=postgres PGUSER=postgres PGPASSWORD="$(cat /etc/yunohost/psql)" psql --dbname="$db" --command="ALTER DATABASE $db REFRESH COLLATION VERSION;"
> done
> ```
> 
> sorry I'm not able to translate that in python language

@Salamandar do you think we should add this to the migration script?

What about the `ignored_python_apps`?

Salamandar committed:
migrations: Add migration script for fixing Python venv after trixie migration

Salamandar commented:
> @Salamandar do you think we should add this to the migration script?

I guess so although i don't understand why the /etc/postgresql-common/user_clusters file had to be edited.

> What about the ignored_python_apps?

hmmmm idk. Also I would like to re-try just upgrading the venv without recreating it completely.

Salamandar committed:
migrations: postgresql: Remove postgresql-common config set by Immich

Immich sets the default version of psql to 15 because they install 17.
Now that Trixie has 17, remove the line.

Signed-off-by: Félix Piédallu <felix@piedallu.me>

Salamandar committed:
migrations: postgresql: run REINDEX DATABASE, REFRESH COLLATION VERSION

See recommendations on https://www.postgresql.org/docs/17/sql-altercollation.html#SQL-ALTERCOLLATION-NOTES

ewilly commented:
> I guess so although i don't understand why the /etc/postgresql-common/user_clusters file had to be edited.

@Salamandar : humm you are right, that's due to immich package. On bookworm I use postgreql17 coming from postgresql repo. To ensure that the default cluster stick to debian one (15 in bookwrom) I fill that file (see https://github.com/YunoHost-Apps/immich_ynh/blob/0b045730eb1e999cff149d7c2089d073ee217476/scripts/_common.sh#L461). With trixie we could go back to because debian provide postgreql17 as default. But for now I am not sure what's the best way how to proceed.

Salamandar commented:
> > I guess so although i don't understand why the /etc/postgresql-common/user_clusters file had to be edited.
> 
> @Salamandar : humm you are right, that's due to immich package. On bookworm I use postgreql17 coming from postgresql repo. To ensure that the default cluster stick to debian one (15 in bookwrom) I fill that file (see [YunoHost-Apps/immich_ynh@`0b04573`/scripts/_common.sh#L461](https://github.com/YunoHost-Apps/immich_ynh/blob/0b045730eb1e999cff149d7c2089d073ee217476/scripts/_common.sh#L461)). With trixie we could go back to because debian provide postgreql17 as default. But for now I am not sure what's the best way how to proceed.

I think the best is to just remove the line `* * 15 main postgres`. Other lines shall stay I think.
